### PR TITLE
transforms: add restrict flag to StencilShapeMinimize pass

### PR DESCRIPTION
**Pull request by emmau678**
Add a restrict flag to the StencilShapeMinimize pass which restricts the grid size of a stencil computation.

### CODE REVIEW

**Comment by superlopuh** [comment]
Could you please complete the checklist in the description?

File: xdsl/transforms/stencil_shape_minimize.py
Comment (n-io) [inline xdsl/transforms/stencil_shape_minimize.py:79]:
```suggestion
    restrict: tuple[int, ...] | None = None
```
This would set a default value to not restrict the size.

File: xdsl/transforms/stencil_shape_minimize.py
Comment (n-io) [inline xdsl/transforms/stencil_shape_minimize.py:86]:
```suggestion
        if restrict:
            PatternRewriteWalker(
                [
                    InvalidateTemps(),
                    RestrictStoreOp(restrict=self.restrict),
                ],
                apply_recursively=False,
            ).rewrite_module(op)
```
This might be able to work in one pass (not sure though if this needs recursive application). There might also be a `shape-inference` pass inside the conditional, since this invalidates temps.

File: xdsl/transforms/stencil_shape_minimize.py
Comment (n-io) [inline xdsl/transforms/stencil_shape_minimize.py:55]: This class might potentially always have a `restrict`, since it can't do much otherwise.

File: xdsl/transforms/stencil_shape_minimize.py
Comment (n-io) [inline xdsl/transforms/stencil_shape_minimize.py:65]: This looks really good. It could probably be a simple and beautiful tuple-comprehension.

File: xdsl/transforms/stencil_shape_minimize.py
Comment (emmau678) [reply xdsl/transforms/stencil_shape_minimize.py:55]: that makes sense, I'll remove the default None value. thanks!

File: xdsl/transforms/stencil_shape_minimize.py
Comment (emmau678) [reply xdsl/transforms/stencil_shape_minimize.py]: this code gives a few errors in the IDE:
1. "restrict" is not defined" -> I'll change it to self.restrict
2. 
```
Argument of type "list[InvalidateTemps | RestrictStoreOp]" cannot be assigned to parameter "pattern" of type "RewritePattern" in function "__init__"
  "list[InvalidateTemps | RestrictStoreOp]" is not assignable to "RewritePattern"Pylance[reportArgumentType](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportArgumentType)
```
```
Argument of type "list[InvalidateTemps | RestrictStoreOp]" cannot be assigned to parameter "pattern" of type "RewritePattern" in function "__init__"
  "list[InvalidateTemps | RestrictStoreOp]" is not assignable to "RewritePattern"Pylance[reportArgumentType](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportArgumentType)
```
-> looking at other examples, if I add in an enclosed GreedyRewritePatternApplier and remove the ```apply_recursively=False```, it seems to get rid of this error but I don't fully understand it - I'll add it into a new commit and remove if necessary

File: xdsl/transforms/stencil_shape_minimize.py
Comment (emmau678) [reply xdsl/transforms/stencil_shape_minimize.py:65]: thanks! I'll attempt to do this

File: xdsl/transforms/stencil_shape_minimize.py
Comment (n-io) [reply xdsl/transforms/stencil_shape_minimize.py]: Yes, apologies!

File: xdsl/transforms/stencil_shape_minimize.py
Comment (n-io) [inline xdsl/transforms/stencil_shape_minimize.py:59]:
```suggestion
```
This is no longer needed

File: xdsl/transforms/stencil_shape_minimize.py
Comment (n-io) [inline xdsl/transforms/stencil_shape_minimize.py:67]: Assigning to `op.bounds` would actually mean the changes pass the rewriter. It may be safer to rebuild the op via `rewriter.replace_matched_op(stencil.StoreOp.get(...))`.

To prevent this from infinitely rewriting the op, it also needs either `apply_recursively=False` in the PatternRewriteWalker, or a check in this class whether this should be applied, for instance, by checking if `new_bounds_attr != op.bounds` to know if this needs updating, and bail out early otherwise.

**Review by n-io — COMMENTED** [review]
Great, while there are ways to combine this into one filecheck mlir file (if you grep for `--check-prefix`), you could also make a new one. Happy to help with either way.

File: xdsl/transforms/stencil_shape_minimize.py
Comment (emmau678) [reply xdsl/transforms/stencil_shape_minimize.py:66]: ok, thanks for that. I'll work on this now

Comment (emmau678) [reply xdsl/transforms/stencil_shape_minimize.py:66]: I think that because I now have GreedyRewritePatternApplier, I can't have apply_recursively=False.. so I will go with the other option for now but please let me know if I'm missing something simpler

**Comment by n-io** [comment]
> > Great, while there are ways to combine this into one filecheck mlir file (if you grep for `--check-prefix`), you could also make a new one. Happy to help with either way.
> 
> any help on how to get started on this would be greatly appreciated!

I just pushed a filecheck to your branch, but it will likely break. The last thing to do is to run shape inference (`ShapeInferencePass().apply(ctx, op)`) inside the `if self.restrict` conditional to tie everything together. I'll let you do the honours :)

File: xdsl/transforms/stencil_shape_minimize.py
Comment (superlopuh) [inline xdsl/transforms/stencil_shape_minimize.py:95]: I would recommend against this

Comment (superlopuh) [reply xdsl/transforms/stencil_shape_minimize.py:95]: @n-io, what's the motivation for doing this inside the pass and not as part of the pass pipeline?

Comment (n-io) [reply xdsl/transforms/stencil_shape_minimize.py:95]: The most immediate purpose is to undo the `InvalidateTemps()` rewrite pattern. The pass in general is for testing purposes, to allow restricting the stencil grid size to something small enough that for instance could run in a simulator. The idea is simple: invalidate temps, modify size on store ops, un-invalidate temps. This needs to happen before `stencil-shape-minimize` is run in order for that to work. The only alternative would be to separate this out from shape-minimize altogether, though running `stencil-shape-minimize{restrict=30,30,30}` seems to make some structural sense in my head instead of having it as a fully independent pass.

Comment (superlopuh) [reply xdsl/transforms/stencil_shape_minimize.py:95]: Ah OK, that makes sense. The main thing that triggered me was calling a pass inside of another pass, for the same reason, as the contract is that a pass leaves the IR in a valid state, if the IR is in a valid state to begin with. In that sense it would make sense to split out the body of the shape inference pass into a helper method, and to call into that from here.

Comment (n-io) [reply xdsl/transforms/stencil_shape_minimize.py:95]: I've seen this done in a few places, though was never sure why helper methods are preferred, including by `post_walk_func`?

Comment (superlopuh) [reply xdsl/transforms/stencil_shape_minimize.py:95]: My understanding is that it's preferable to expose the least restrictive API when possible. For example, there's no need to pass the context when performing shape inference. It's probably also to make things more flexible in general, like you may want to only perform shape inference on a subset of your module.

Comment (emmau678) [reply xdsl/transforms/stencil_shape_minimize.py]: I'm not sure I follow - do I need to make some updates here?

Comment (superlopuh) [reply xdsl/transforms/stencil_shape_minimize.py]: Yes, please! Could pull out the body of the shape inference pass into a python function that takes any operation, and no MLContext, and call that helper function from here instead of the pass?

File: xdsl/transforms/shape_inference.py
Comment (superlopuh) [inline xdsl/transforms/shape_inference.py:39]: Our coding style is to use snake_case for function names, in this case giving something like this:
```suggestion
def shape_inference_pass_helper(op: builtin.ModuleOp):
```
Could you please move this above the pass definition, and also call it from the pass itself?

**Review by superlopuh — APPROVED** [review]
It would probably be good to get @n-io's tick before merging, but LGTM!

File: xdsl/transforms/stencil_shape_minimize.py
Comment (n-io) [inline xdsl/transforms/stencil_shape_minimize.py:65]: nit: An alternative solution that didn't come to mind earlier would be to build (or pass in) a `restrict_bounds` StencilBoundsAttr and use one of its overload, i.e. `op.bounds & restrict_bounds` to build the new bounds. Feel free to ignore, but wanted to mention.

**Review by n-io — APPROVED** [review]
LGTM, feel free to merge!

File: xdsl/transforms/stencil_shape_minimize.py
Comment (emmau678) [reply xdsl/transforms/stencil_shape_minimize.py:65]: noted, thanks for this

**Pull request merged**
